api/src/main/java/net/slimediamond/espial/api/action/BlockAction.java

```diff
@@ -134,7 +134,7 @@ public BlockAction build() {
         @Override
         public SubmittableResult<BlockRecord> submit() throws Exception {
           return (SubmittableResult<BlockRecord>)
-              Espial.getInstance().getEspialService().submitAction(this);
+            EspialProviders.getEspialService().submitAction(this);
```

Comment:
Saw in this file you have ``location`` method in the builder. Save yourself a line all over and take the world from the location, but also add one with ``Vector3i`` as Sponge uses Vector3i as a blocks position. This would then allow you to use snapshots without assuming the Location object is present (something you should never do). 

So essentially

```java

public Builder location(ServerLocation loc){
  this.x = loc.x();
....
this.worldKey = loc.world().key();
} 

public Builder location(Vector3i position){
  this.x = position.x();
 ...
}

public Builder snapshot(BlockSnapshot snapshot){
  return position(snapshot.position()).world(snapshot.worldKey());
}

```

Reply:
A good idea! I've implemented now. Thanks.